### PR TITLE
Support workflows without user interaction

### PR DESCRIPTION
**Pull request by lmm-git**
This PR introduces multiple new features and minor fixes in order to allow `oidc-agent` to be used programmatically by other (wrapper) programs.

We have the use case of setting up `oidc-agent` during login for a specific user, meaning a PAM module will set up the `oidc-agent` with a (refresh) token provider, which then can used by the user via `oidc-token`. As this setup is done on every login/PAM invocation, we e.g. do not need to store the accounts on disk.

When starting the `oidc-agent`, we pass the generated environment vars to the users environment by utilizing `pam_putenv` in a custom PAM module. That's why we implemented a json return for those variables to be able to parse them by another program and pass it to the envs of the user later.

As we are using this stack on Ubuntu 21.04, we also fixed compilation for those newer OS, which enables new compiler warnings by default.

I believe this PR is reviewed best by going through each commit. If you have any questions I am happy to answer them.

### CODE REVIEW

**Review by zachmann — COMMENTED** [review]
Hi Leonard,

many thanks for the PR. We appreciate your contribution.

I have some comments/requests for changes. I added some review comments, but these are mainly for explanation. You don't have to implement these. Since most are only small and easy changes I thought it would be easier if I do it myself.

I submitted a PR back with my changes. Feel free to go through them and merge them.
I then can merge this PR.

File: src/utils/printerUtils.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    sprintf(pid_str, "%d", daemon_pid);
+
+    // print empty string of daemon_socket is null
+    char *daemon_socket_str = daemon_socket;
+    if (daemon_socket_str == NULL) {
+      daemon_socket_str = "";
+    }
+    
+    cJSON *outer_json = generateJSONObject("socket", cJSON_String, daemon_socket_str, "dpid", cJSON_String, pid_str, NULL);
+    printStdout("%s", jsonToString(outer_json));
```
[the remark this answers is not in the document]
Comment (lmm-git) [reply src/utils/printerUtils.c:31]: Uff, you are right. Obviously I wasn't concentrated when writing this code..

File: src/oidc-gen/oidc-gen_options.c [hunk 2]
```diff
@@ -333,6 +371,20 @@ static error_t parse_opt(int key, char* arg, struct argp_state* state) {
...
+        exit(EXIT_FAILURE);
+      }
+      // Copy env_pass as subsequent getenv calls might modify our just received data
+      arguments->refresh_token = oidc_strcopy(env_refresh_token);
+      if (arguments->flows == NULL) {
+        arguments->flows        = list_new();
+        arguments->flows->match = (matchFunction)strequal;
+      }
+      list_rpush(arguments->flows, list_node_new("refresh"));
+      break;
```
[the remark this answers is not in the document]
Comment (lmm-git) [reply src/oidc-gen/oidc-gen_options.c:387]: Thought about this as well but as the other envs were not configurable I thought I will stick to it. Like it now much more :)

**Comment by lmm-git** [comment]
Thanks for going through the changes!

Will check your changes and verify they still work as expected soon!

**Comment by lmm-git** [comment]
Just checked you MR, lgtm!

All basic functions seem still to work correctly :)

Thank you very much for your effort!

**Pull request merged**